flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/ExecutingTest.java

```diff
@@ -0,0 +1,444 @@
+public class ExecutingTest extends TestLogger {
+    @Test
+    public void testTransitionToFailing() throws Exception {
+        final String failureMsg = "test exception";
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectFailing(
+                    (failingArguments -> {
+                        assertThat(failingArguments.getExecutionGraph(), notNullValue());
+                        assertThat(failingArguments.getFailureCause().getMessage(), is(failureMsg));
+                    }));
+            ctx.setHowToHandleFailure(Executing.FailureResult::canNotRestart);
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            exec.handleGlobalFailure(new RuntimeException(failureMsg));
+        }
+    }
+
+    @Test
+    public void testTransitionToRestarting() throws Exception {
+        final Duration duration = Duration.ZERO;
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectRestarting(
+                    (restartingArguments ->
+                            assertThat(restartingArguments.getBackoffTime(), is(duration))));
+            ctx.setHowToHandleFailure((t) -> Executing.FailureResult.canRestart(duration));
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            exec.handleGlobalFailure(new RuntimeException("Recoverable error"));
+        }
+    }
+
+    @Test
+    public void testTransitionToCancelling() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectCancelling(assertNonNull());
+            Executing exec = getExecutingState(ctx);
+            exec.onEnter();
+            exec.cancel();
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedOnTerminalState() throws Exception {
+        ExecutorService executor = Executors.newSingleThreadExecutor();
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectFinished(
+                    archivedExecutionGraph ->
+                            assertThat(archivedExecutionGraph.getState(), is(JobStatus.FAILED)));
+            ctx.setGetMainThreadExecutor(() -> executor);
+            Executing exec = getExecutingState(ctx);
+            ctx.setExpectedStateChecker((state) -> state == exec);
+            exec.onEnter();
+            // transition EG into terminal state, which will notify the Executing state about the
+            // failure (async via the supplied executor)
+            exec.getExecutionGraph().failJob(new RuntimeException("test failure"));
+            executor.shutdown();
+            executor.awaitTermination(10, TimeUnit.SECONDS);
+        }
+    }
+
+    @Test
+    public void testTransitionToFinishedOnSuspend() throws Exception {
+        try (MockExecutingContext ctx = new MockExecutingContext()) {
+            ctx.setExpectFinished(
```

Comment:
hmm...now that I see it in action I'm wondering...like, we _could_ just capture the state transition in the context and then do an assert on it, right? We don't actually need all this trap business that I proposed.

Structurally, the test would then look like this, which is more in line with our usual test patterns:
```
try (MockExecutingContext ctx = new MockExecutingContext()) {
    Executing exec = getExecutingState(ctx);
    exec.onEnter();
    exec.suspend(new RuntimeException("suspend"));

    ctx.assertFinished(
            archivedExecutionGraph -> {
                assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED));
            });
}
```
where the assert fails immediately if the transition did not occur (i.e., the *Arguments are null).

Reply:
This is a bit more intuitive to read, yes. But I believe it leads to longer test methods (because we always need to assert something state related in the end (I don't think using some fancy logic in AutoClosable.close() to decide if no or only one asserted state transition happened makes sense))

The current (PR) approach has the benefit that it fails if an unexpected state transition is happening. We would need to introduce the methods `ctx.assertOnlyFinished(..)` and  `ctx.assertNoStateTransition()`.

Also, I don't think the implementation of the context would *really* benefit, might actually be more tedious:
![image](https://user-images.githubusercontent.com/89049/106745604-a39f1d80-6621-11eb-8253-e5cd6ef5d251.png)

I'm leaning towards leaving it as-is.

Reply:
> because we always need to assert something state related in the end

I don't think we'd need to; as-is if we don't setup an expectation then a state transition might occur without the test failing. Why would we always need an assertion now?

> [we'd need] ctx.assertOnlyFinished(..)

We should in general enforce that only 1 state transition is triggered. If this happens it is either a bug in the state implementation or in the test itself.

> [we'd need] ctx.assertNoStateTransition()

I'd think we'll need that at some point anyway.

> context [...] might actually be more tedious:

You'd only need a shared helper method checking that only 1 of the arguments is not-null. Coupled with the initial assertion this ensures that only the transition in question has occurred.
Not even that; you just need a boolean flag that is set if the state transition.

Reply:
Okay, I migrated the ExecutingTests to the new model.